src/tests.rs

```diff
@@ -0,0 +1,176 @@
+    sync::atomic::{AtomicU32, Ordering},
+    vec,
+};
+
+#[cfg(all(
+    any(target_arch = "wasm32", target_arch = "wasm64"),
+    target_os = "unknown",
+))]
+pub(crate) use wasm_bindgen_test::wasm_bindgen_test as test;
+
+#[cfg(feature = "test-in-browser")]
+wasm_bindgen_test::wasm_bindgen_test_configure!(run_in_browser);
+
+fn num_diff_bits(s1: &[u8], s2: &[u8]) -> usize {
+    assert_eq!(s1.len(), s2.len());
+    s1.iter()
+        .zip(s2.iter())
+        .map(|(a, b)| (a ^ b).count_ones() as usize)
+        .sum()
+}
+
+// For calls of size `len`, count the number of bits which differ between calls
+// and check that between 3 and 5 bits per byte differ. Probability of failure:
+// ~ 10^(-30) = 2 * CDF[BinomialDistribution[8*256, 0.5], 3*256]
+fn check_bits<T: Copy>(len: usize, v: T, f: fn(&mut [T]) -> Result<&mut [u8], Error>) {
+    let mut num_bytes = 0;
+    let mut diff_bits = 0;
+    while num_bytes < 256 {
+        let mut v1 = vec![v; len];
+        let s1 = f(&mut v1).unwrap();
+        let mut v2 = vec![v; len];
+        let s2 = f(&mut v2).unwrap();
+
+        num_bytes += len;
+        diff_bits += num_diff_bits(s1, s2);
+    }
+
+    assert!(diff_bits > 3 * num_bytes);
+    assert!(diff_bits < 5 * num_bytes);
+}
+
+pub(crate) fn check_small<T: Copy>(v: T, f: fn(&mut [T]) -> Result<&mut [u8], Error>) {
+    for len in 1..=64 {
+        check_bits(len, v, f)
+    }
+}
+
+fn wrapped_fill(dest: &mut [u8]) -> Result<&mut [u8], Error> {
+    getrandom(dest).map(|()| dest)
+}
+
+#[test]
+fn fill_zero() {
+    getrandom(&mut []).unwrap();
+}
+#[test]
+fn fill_zero_uninit() {
+    getrandom_uninit(&mut []).unwrap();
+}
+#[test]
+fn fill_small() {
+    check_small(0, wrapped_fill);
+}
+#[test]
+fn fill_small_uninit() {
+    check_small(MaybeUninit::uninit(), getrandom_uninit);
+}
+#[test]
+fn fill_large() {
+    check_bits(1_000, 0, wrapped_fill);
+}
+#[test]
+fn fill_large_uninit() {
+    check_bits(1_000, MaybeUninit::uninit(), getrandom_uninit);
+}
+#[test]
+fn fill_huge() {
+    check_bits(1_000_000, 0, wrapped_fill);
+}
+#[test]
+fn fill_huge_uninit() {
+    check_bits(1_000_000, MaybeUninit::uninit(), getrandom_uninit);
+}
+
+// On WASM, the thread API always fails/panics
+#[cfg(not(any(target_arch = "wasm32", target_arch = "wasm64")))]
+#[test]
+fn multithreading() {
+    use std::{sync::mpsc, thread};
+
+    let mut txs = vec![];
+    for _ in 0..20 {
+        let (tx, rx) = mpsc::channel();
+        txs.push(tx);
+
+        thread::spawn(move || {
+            // wait until all the tasks are ready to go.
+            rx.recv().unwrap();
+            let mut v = [0u8; 1000];
+
+            for _ in 0..100 {
+                getrandom(&mut v).unwrap();
+                thread::yield_now();
+            }
+        });
+    }
+
+    // start all the tasks
+    for tx in txs.iter() {
+        tx.send(()).unwrap();
+    }
+}
+
+// A simple Linear Congruential Generator (LCG) that passes our RNG tests.
+const A: u32 = 134775813;
```

Comment:
I don't think we should spend effort on a custom implementation that passes the tests. The fact that this implementation passes the current test suite is a defect in the test suite that can and should be fixed.

Also, we're not testing a custom implementation; we're testing the mechanics of the custom implementation registration and how we deal with the unique situations that a custom implementation gives us.

What we'll likely need is a mock implementation that we can programatically control the behavior of.

Reply:
This is a good point, I agree. The RNG quality tests now don't run if the `"custom"` feature is enabled. Other changes to the testing of the custom RNG are moved to #473.